### PR TITLE
[GLUTEN-4398][IT] Add Golden Files for TPC-H Spark32 + Gluten Execution Plan 

### PR DESCRIPTION
**Pull request by zwangsheng**
## What changes were proposed in this pull request?

Add tpc-h golden file for spark32 and check whether PR changed Spark Execution Plan

Continue worker with #4399

1. Add golden files for tpc-h(except delta related)
2. for now only work for spark32, other spark version will be made later.
3. Each PR will check those golden files though Velox Backend Unit Test, and will upload actual golden files to github action if some query mis-match.

截屏2024-02-19 16 35 31

    
Users can download and replace the local corresponding file for quick correction.
Golden files located at `${GLUTEN_REPO_HOME}/backends-velox/src/test/resources/tpch-approved-plan/${TPC_H_TEST_SUB_TYPE}/${SPARK_VERSION}/${QUERY_ID}.txt`, 
for example: 
`${GLUTEN_REPO_HOME}/backends-velox/src/test/resources/tpch-approved-plan/gluten-bhj-vanilla-be/spark322/1.txt`.

And when user run unit test in local, them can find actual golden file in tmp dir, find this with unit test output log.

```log
- TPC-H q5 *** FAILED ***
  Mismatch for query 5
  Actual Plan path: /tmp/tpch-approved-plan/v2-bhj/spark322/5.txt
  Golden Plan path: /opt/gluten/backends-velox/target/scala-2.12/test-classes/tpch-approved-plan/v2-bhj/spark322/5.txt (VeloxTPCHSuite.scala:101)
```

## How was this patch tested?

Add Unit Test for Spark32 to check golden files

### CODE REVIEW

**Comment by zhztheplayer** [comment]
@zwangsheng Thanks in advance.

Would you like to introduce a bit about the term "golden file"? Thanks.

If I understand it correctly, when execution plan for a query gets changed, developer should regenerate the files using `gluten-it`, is that correct?

If yes, I am worried that this way puts extra burden to developers since compiling `gluten-it` takes quiet a few time as it packages everything. Did you consider adding this check to Gluten's main code?

After all, would you like to share a bit about the background why you wanted to add this check to Gluten (and its CI)? Thanks.

**Comment by zwangsheng** [comment]
Thanks for noticing this feature. @zhztheplayer 

> Would you like to introduce a bit about the term "golden file"? Thanks.

Base on Spark SQL Module has [PlanStabilitySuite](https://github.com/apache/spark/blob/da0c31c629ca378704cb7f97c9983be9b6ca96c4/sql/core/src/test/scala/org/apache/spark/sql/PlanStabilitySuite.scala#L72) to sense how code changes (from PR) change the execution plan.

Gluten is closely related to spark execution plans, and `Golden File` should be introduced for gluten to ensure that code changes to the execution plan are the first to be detected, so that we can better evaluate code changes.

> If I understand it correctly, when execution plan for a query gets changed, developer should regenerate the files using gluten-it, is that correct?

Yes, if devs make same change to related code, the Golden Files should be re-generate by `gluten-it` module

> If yes, I am worried that this way puts extra burden to developers since compiling gluten-it takes quite some time as it packages everything.

At present, it seems that the process of user contribution PR will be affected, and maybe we can make some changes based on this PR to reduce the compilation cost of the user side(in the process of CI detection, we can upload the changed Golden File to the log store section of Github Action, and users can download it instead of compiling it locally, that seems to allay the concerns you mentioned)

But in general, we still should have such a Golden File mechanism in case we miss the relevant changes.

Thanks again, if have new ideas to help us check spark execution plan changes, be open to them.

**Comment by zhztheplayer** [comment]
Thanks for the detailed explanation @zwangsheng.

> At present, it seems that the process of user contribution PR will be affected, and maybe we can make some changes based on this PR to reduce the compilation cost of the user side(in the process of CI detection, we can upload the changed Golden File to the log store section of Github Action, and users can download it instead of compiling it locally, that seems to allay the concerns you mentioned)

I still feel the procedure is probably a little too complicated for developers. Gluten has in-place TPC-H unit test code here

https://github.com/oap-project/gluten/blob/main/backends-velox/src/test/scala/io/glutenproject/execution/VeloxTPCHSuite.scala

So if possible I might be inclined to have the golden checking integrated with current unit tests rather than integration testing. Then developers could avoid re-building the whole project to change golden files when plan gets changed. I understand it might require for a bunch of changes to this PR but let's see if it is worth it.

I was considering bring some `gluten-it`'s features into unit tests but that may mess up the responsibility between ut and it. I am not sure. But currently the cost of using `gluten-it` to manually verify code changes is high.

A simple comparison between using it and ut:

it:

Developer change code ->  Commits code change -> Run GHA CI -> Download golden files -> Replace local golden files -> Rerun GHA CI

ut (probably):
Developer change code -> Run UT to generate new golden files -> Replace local golden files -> Commit code change -> Run GHA CI

I assume the ut solution should take much shorter time than using it?

**Comment by zwangsheng** [comment]
> ut (probably):
Developer change code -> Run UT to generate new golden files -> Replace local golden files -> Commit code change -> Run GHA CI

Thanks, i will give a try to move those golden files test to unit test.

**Comment by zhztheplayer** [comment]
> > ut (probably):
> > Developer change code -> Run UT to generate new golden files -> Replace local golden files -> Commit code change -> Run GHA CI
> 
> Thanks, i will give a try to move those golden files test to unit test.

Thank you!

**Comment by zwangsheng** [comment]
Hi @zhztheplayer, sorry for late reply. Here move golden file check from IT to UT modules.

And for now, add golden files for all gluten + spark32 golden files check.

**Comment by zhztheplayer** [comment]
> Hi @zhztheplayer, sorry for late reply. Here move golden file check from IT to UT modules.
> 
> And for now, add golden files for all gluten + spark32 golden files check.
> 
> I will add upload CI to upload changed golden files to github action, for devs to download. Before this, please take your free time to review this golden files and unit test. Thanks 😄 .

Thank you for the update. And would you like to fix PR description to describe how to re-generate golden files locally from user side now? Also we can add comment for same purpose to `VeloxTPCHSuite.scala`. Thanks.

**Comment by zwangsheng** [comment]
> Thank you for the update. And would you like to fix PR description to describe how to re-generate golden files locally from user side now? Also we can add comment for same purpose to `VeloxTPCHSuite.scala`. Thanks.

Yeah, i will finish this up and given that we are currently on the holidays, progress may be slow.

File: tools/gluten-te/centos/gha/gha-checkout/cp.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -ex
+
+if [ -z "$GITHUB_RUN_ID" ]
+then
+  echo "Unable to parse GITHUB_RUN_ID."
+  exit 1
+fi
```
Comment (zhztheplayer) [inline tools/gluten-te/centos/gha/gha-checkout/cp.sh:23]: Should we check `$GITHUB_JOB` too?

**Review by zhztheplayer — COMMENTED** [review]
Overall looks good. Thanks @zwangsheng !

**Review by ulysses-you — COMMENTED** [review]
it seems overkill to generate golden files for each type 
- gluten-bhj-vanilla-be
- vanilla-bhj-gluten-be
- partitioned
- v1
- v1-bhj
- v2
- v2-bhj

Can we just maintain `v1` and `v1-bhj` ?

**Comment by zwangsheng** [comment]
> it seems overkill to generate golden files for each type
> 
> * gluten-bhj-vanilla-be
> * vanilla-bhj-gluten-be
> * partitioned
> * v1
> * v1-bhj
> * v2
> * v2-bhj
> 
> Can we just maintain `v1` and `v1-bhj` ?

Reducing type seems feasible, but considering Spark retains v2 related, should we also retain v2 related?

**Comment by ulysses-you** [comment]
I think no body would use v2 code path to read parquet.. it is mainly used for data lake.

**Comment by zhztheplayer** [comment]
Another question: After this patch, is there any best practice to re-generate all the golden files through a single command?

>
>```
>- TPC-H q5 *** FAILED ***
>  Mismatch for query 5
>  Actual Plan path: /tmp/tpch-approved-plan/v2-bhj/spark322/5.txt
>  Golden Plan path: /opt/gluten/backends-velox/target/scala-2.12/test-classes/tpch-approved-plan/v2-bhj/spark322/5.txt (VeloxTPCHSuite.scala:101)
>```

The example is good but it seems developer should fix the files one by one?

**Comment by zwangsheng** [comment]
> Another question: After this patch, is there any best practice to re-generate all the golden files through a single command?
> 
> > ```
> > - TPC-H q5 *** FAILED ***
> >  Mismatch for query 5
> >  Actual Plan path: /tmp/tpch-approved-plan/v2-bhj/spark322/5.txt
> >  Golden Plan path: /opt/gluten/backends-velox/target/scala-2.12/test-classes/tpch-approved-plan/v2-bhj/spark322/5.txt (VeloxTPCHSuite.scala:101)
> > ```
> 
> The example is good but it seems developer should fix the files one by one?

Nope, github action will cache those actual golden files in tgz mode, so users can download and replace all in one click.

**Comment by ulysses-you** [comment]
lgtm, please add a step by step docs to help contributors check and update golden files.

**Comment by zwangsheng** [comment]
Hi @ulysses-you @zhztheplayer Golden files related introduction was added in `NewToGluten.md`, please help to correct it

File: docs/developers/NewToGluten.md [hunk 2]
```diff
@@ -374,3 +374,20 @@ spark-shell --name run_gluten \
...
+and this will be triggered when running the unit test.
+
+As a result, developers may encounter unit test fail in Github CI or locally, with the following error message:
+```log
+- TPC-H q5 *** FAILED ***
+  Mismatch for query 5
+  Actual Plan path: /tmp/tpch-approved-plan/v2-bhj/spark322/5.txt
+  Golden Plan path: /opt/gluten/backends-velox/target/scala-2.12/test-classes/tpch-approved-plan/v2-bhj/spark322/5.txt (VeloxTPCHSuite.scala:101)
+```
+For developers to update the golden plan, you can find the actual plan in Github CI Artifacts or in local `/tmp/` directory.
```
Comment (ulysses-you) [inline docs/developers/NewToGluten.md:391]: can we add a `.png` to make it clear ?

**Review by ulysses-you — APPROVED** [review]
lgtm, cc @zhztheplayer @PHILO-HE if you have other comments

**Review by zhztheplayer — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**